### PR TITLE
Add `ParametersEncoder` protocol and `URLParametersEncoder` type.

### LINKED ISSUE #11: Create `URLParametersEncoder`
### This task is part of
- https://github.com/ios-course/simple-network-service/issues/6

### Depends on tasks:
 - https://github.com/ios-course/simple-network-service/issues/3

### Possible implmentation

``` swift
public protocol ParametersEncoder {
    static func encode(_ parameters: URLParameters?, into request: inout URLRequest) // TODO: do we need throws?
}

public struct URLParametersEncoder: ParametersEncoder {
    public static func encode(_ parameters: URLParameters?, into request: inout URLRequest) {
        guard let url = request.url, let parameters else { return }

        var urlComponents = URLComponents(url: url, resolvingAgainstBaseURL: false)
        urlComponents?.queryItems = queryItems(from: parameters)

        request.url = urlComponents?.url
    }

    private static func queryItems(from parameters: URLParameters) -> [URLQueryItem] {
        parameters.map { key, value in
            URLQueryItem(
                name: key,
                value: "\(value)".addingPercentEncoding(withAllowedCharacters: .urlPathAllowed)
            )
        }
    }
}
```

### PR DESCRIPTION
**Pull request by kirpichonok**
### Summary
This PR is a part of the work to resolve #11 .

### Changes
- Add `ParametersEncoder` protocol.
- Add `ParametersEncoder` type implementation.
- Add unit tests.

### Checks to complete

- [x] The PR title has a conventional commit style (short and descriptive).
- [x] The PR description includes essential information that helps understand the purpose of the changes.
- [x] I've built and run my changes, and no warnings or errors occurred.
- [x] All existing tests passed with my changes.
- [x] My code follows our style guide.
- [x] I've performed a self-review of my code.
- [x] Every non-private interface is documented properly.
- [x] I've added tests for my code (if it makes sense).
- [x] I've updated `CHANGELOG.md` file.
- [x] PR has assignee, reviewers, milestone, it is properly linked to the project and to the issue.

### CODE REVIEW

File: Sources/NetworkService/ParametersEncoder/URLParametersEncoder.swift [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import Foundation
+
+public struct URLParametersEncoder: ParametersEncoder {
```
Comment (NadzeyaShpakouskaya) [inline Sources/NetworkService/ParametersEncoder/URLParametersEncoder.swift]: missing doc

File: Tests/NetworkServiceTests/URLParametersEncoderTests.swift [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+        XCTAssertEqual(actualQueryItems, expectedQueryItems)
+    }
+
+    private func queryItems(from parameters: URLParameters?) -> [URLQueryItem]? {
+        guard let parameters else { return nil }
+        return parameters.map { key, value in
+            URLQueryItem(name: key, value: value)
+        }
+    }
```
Comment (NadzeyaShpakouskaya) [inline Tests/NetworkServiceTests/URLParametersEncoderTests.swift]: Does it make sense to make this function internal in NetworkService instead of duplicating it in tests? I'm ok with your implementation as well.

Do we need an optional array as a return value? I think an empty array is ok for our case.

File: Tests/NetworkServiceTests/URLParametersEncoderTests.swift [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    private func assertQueryItemsEquality(parameters: URLParameters?) throws {
+        let expectedQueryItems = queryItems(from: parameters)
+
+        URLParametersEncoder.encode(parameters, into: &request)
+        let components = try XCTUnwrap(URLComponents(url: request.url!, resolvingAgainstBaseURL: false))
+        let actualQueryItems = components.queryItems
+
+        XCTAssertEqual(actualQueryItems, expectedQueryItems)
+    }
+
```
Comment (NadzeyaShpakouskaya) [inline Tests/NetworkServiceTests/URLParametersEncoderTests.swift]: I don't think that creating a separate function is a good idea. I know that it looks like duplicating code for every test method, but for me, it's fine for test functions. 
I prefer more robust and independent test cases. If the test fails, you should jump into the implementation of `assertQueryItemsEquality`.

File: Tests/NetworkServiceTests/URLParametersEncoderTests.swift [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    // MARK: - Private interface
+
+    private func assertQueryItemsEquality(parameters: URLParameters?) throws {
+        let expectedQueryItems = queryItems(from: parameters)
+
+        URLParametersEncoder.encode(parameters, into: &request)
+        let components = try XCTUnwrap(URLComponents(url: request.url!, resolvingAgainstBaseURL: false))
+        let actualQueryItems = components.queryItems
+
+        XCTAssertEqual(actualQueryItems, expectedQueryItems)
```
Comment (NadzeyaShpakouskaya) [inline Tests/NetworkServiceTests/URLParametersEncoderTests.swift]: Can we simplify it.
```suggestion
        XCTAssertEqual( components.queryItems, expectedQueryItems)
```

File: Tests/NetworkServiceTests/URLParametersEncoderTests.swift [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    override func tearDown() {
+        super.tearDown()
+        request = nil
+    }
+
+    func testQueryItemsPropertyOfRequestIsNilWhenNilIsPassedAsParameters() throws {
+        try assertQueryItemsEquality(parameters: nil)
+    }
```
Comment (NadzeyaShpakouskaya) [inline Tests/NetworkServiceTests/URLParametersEncoderTests.swift]: I'm not sure how we should properly test the parameters, maybe we can add more assertions

```swift
    func testQueryItems() throws {
        // Given
        let parameters = [
            "a": "a"
        ]
        
        // When
        URLParametersEncoder.encode(parameters, into: &request)
        
        // Then
        XCTAssertEqual(request?.url?.query,"a=a")
        XCTAssertEqual(request?.url?.absoluteString,"\(testURL.absoluteString)?a=a")
    }

    func testQueryItemsTwo() throws {
        // Given
        let parameters = [
            "a": "a",
            "bc": ""
        ]
        
        // When
        URLParametersEncoder.encode(parameters, into: &request)
        
        // Then
        XCTAssertEqual(request?.url?.query,"a=a&bc=")
        XCTAssertEqual(request?.url?.absoluteString,"\(testURL.absoluteString)?a=a&bc=")
    }
```

File: Sources/NetworkService/ParametersEncoder/ParametersEncoder.swift [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import Foundation
+
+/// A type that encodes URL parameters into a request.
```
Comment (iBYarchitect) [inline Sources/NetworkService/ParametersEncoder/ParametersEncoder.swift:3]:
```suggestion
/// A type responsible for encoding URL parameters into a request.
```

File: Sources/NetworkService/ParametersEncoder/URLParametersEncoder.swift (hunk 1 above)
Comment (iBYarchitect) [inline Sources/NetworkService/ParametersEncoder/URLParametersEncoder.swift:3]: Should you add documentation here?
```suggestion
/// Encodes URL parameters into a request's URL.
public struct URLParametersEncoder: ParametersEncoder {
```

File: Tests/NetworkServiceTests/URLParametersEncoderTests.swift [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+    func testQueryItemsPropertyOfRequestIsEmptyArrayWhenEmptyDictionaryIsPassedAsParameters() throws {
+        try assertQueryItemsEquality(parameters: [:])
+    }
+
+    func testQueryItemsPropertyOfRequestHasExpectedValueWhenPassedOneKeyValuePairOfURLParameters() throws {
```
Comment (iBYarchitect) [inline Tests/NetworkServiceTests/URLParametersEncoderTests.swift]: Maybe you could simplify the name of the tests or the meaning in this case could lost?
```suggestion
    func testQueryItemsPropertyHasSingleItemWhenSingleKeyValuePairPassed() throws {
```

**Review by iBYarchitect — CHANGES_REQUESTED** [review]
Good attempt!
There are a few places for improvement:

File: Sources/NetworkService/ParametersEncoder/ParametersEncoder.swift (hunk 6 above)
Comment (kirpichonok) [reply Sources/NetworkService/ParametersEncoder/ParametersEncoder.swift:3]: I think your version doesn't add clarity.

File: Tests/NetworkServiceTests/URLParametersEncoderTests.swift (hunk 7 above)
Comment (kirpichonok) [reply Tests/NetworkServiceTests/URLParametersEncoderTests.swift]: Your proposal doesn't explain the test's meaning. It doesn't just check the amount of query Items, but it checks their equality to expected values as well.

File: Sources/NetworkService/ParametersEncoder/ParametersEncoder.swift [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+public protocol ParametersEncoder {
+    /// Encodes provided set of parameters as URL parameters into the given request's URL.
+    ///
+    /// - Parameters:
+    ///  - parameters: An optional set of URL parameters to be encoded.
+    ///  - request: A request object to which URL the encoded parameters will be assigned.
```
Comment (NadzeyaShpakouskaya) [inline Sources/NetworkService/ParametersEncoder/ParametersEncoder.swift]: Please check the formatting, the  doc comments don't display properly
Screenshot 2023-08-23 at 09 56 44

```suggestion
    /// - Parameters:
    ///     - parameters: An optional set of URL parameters to be encoded.
    ///     - request: A request object to which URL the encoded parameters will be assigned.
```

**Review by NadzeyaShpakouskaya — CHANGES_REQUESTED** [review]
Well done 👍 
Some suggestion:
 - Resolve merge conflict.
 - Follow the same folder structure in the test target.


Please consider some improvements:

File: Tests/NetworkServiceTests/URLParametersEncoderTests.swift (hunk 4 above)
Comment (kirpichonok) [reply Tests/NetworkServiceTests/URLParametersEncoderTests.swift]: I think it's not clearly enough when we use `components.queryItems`. The additional constant would not harm in this case. 
WDYT?

File: Sources/NetworkService/ParametersEncoder/ParametersEncoder.swift (hunk 8 above)
Comment (kirpichonok) [reply Sources/NetworkService/ParametersEncoder/ParametersEncoder.swift]: Fixed

File: Tests/NetworkServiceTests/URLParametersEncoderTests.swift (hunk 2 above)
Comment (kirpichonok) [reply Tests/NetworkServiceTests/URLParametersEncoderTests.swift]: Fixed.

File: Sources/NetworkService/ParametersEncoder/URLParametersEncoder.swift (hunk 1 above)
Comment (kirpichonok) [reply Sources/NetworkService/ParametersEncoder/URLParametersEncoder.swift]: Fixed.

Comment (kirpichonok) [reply Sources/NetworkService/ParametersEncoder/URLParametersEncoder.swift:3]: Fixed.

File: Tests/NetworkServiceTests/URLParametersEncoderTests.swift (hunk 5 above)
Comment (kirpichonok) [reply Tests/NetworkServiceTests/URLParametersEncoderTests.swift]: We have an additional problem with the order of the parameters.

File: Tests/NetworkServiceTests/URLParametersEncoderTests.swift (hunk 3 above)
Comment (kirpichonok) [reply Tests/NetworkServiceTests/URLParametersEncoderTests.swift]: I've discussed it on the call today and decided to keep this version.

**Review by iBYarchitect — COMMENTED** [review]
LGTM

**Review by iBYarchitect — APPROVED** [review]
LGTM

File: Tests/NetworkServiceTests/URLParametersEncoderTests.swift (hunk 4 above)
Comment (NadzeyaShpakouskaya) [reply Tests/NetworkServiceTests/URLParametersEncoderTests.swift]: You can use a more specific name  for var `components`
```swift
XCTAssertEqual(actualURLComponents.queryItems, expectedQueryItems)
```
or
```swift
XCTAssertEqual(existingURLComponents.queryItems, expectedQueryItems)
```

File: Tests/NetworkServiceTests/URLParametersEncoderTests.swift (hunk 5 above)
Comment (NadzeyaShpakouskaya) [reply Tests/NetworkServiceTests/URLParametersEncoderTests.swift]: Thanks for clarification, you are right

File: Sources/NetworkService/ParametersEncoder/URLParametersEncoder.swift [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    /// - Returns: The collection of resulting query items.
+    /// - Complexity: O(*N*), where *N* is the length of the set of parameters.
+    public static func queryItems(from parameters: URLParameters) -> [URLQueryItem] {
+        parameters.map { key, value in
+            URLQueryItem(
+                name: key,
+                value: value
+            )
+        }
+    }
```
Comment (NadzeyaShpakouskaya) [inline Sources/NetworkService/ParametersEncoder/URLParametersEncoder.swift:33]: Do you think we should add tests for this public function?

**Review by NadzeyaShpakouskaya — CHANGES_REQUESTED** [review]
Almost done, still need some clarifications. 😄

File: Tests/NetworkServiceTests/URLParametersEncoderTests.swift (hunk 3 above)
Comment (NadzeyaShpakouskaya) [reply Tests/NetworkServiceTests/URLParametersEncoderTests.swift]: Nit:
I still prefer a more readable and clear version for the tests.
What do you think about creating a function for extracting queryItems like this
```swift
 private func getQueryItems(from request: URLRequest) -> [URLQueryItem]? {
        let components = try? XCTUnwrap(URLComponents(url: request.url!, resolvingAgainstBaseURL: false))
        
        return components?.queryItems
    }
    
    
    func testQueryItems() throws {
        // Given
        let parameters = ["a": ""]
        
        // When
        URLParametersEncoder.encode(parameters, into: &request)
        let actualQueryItems = getQueryItems(from: request)
        
        // Then
        XCTAssertEqual(actualQueryItems, URLParametersEncoder.queryItems(from: parameters))
    }
```

If you disagree, you can keep your version and resolve the conversation.

File: Sources/NetworkService/ParametersEncoder/URLParametersEncoder.swift (hunk 9 above)
Comment (kirpichonok) [reply Sources/NetworkService/ParametersEncoder/URLParametersEncoder.swift:33]: Yes, sure. Fixed.

File: Sources/NetworkService/ParametersEncoder/URLParametersEncoder.swift [hunk 10]
```diff
@@ -10,7 +18,14 @@ public struct URLParametersEncoder: ParametersEncoder {
         request.url = urlComponents?.url
     }
 
-    private static func queryItems(from parameters: URLParameters) -> [URLQueryItem] {
+    /// Create a collection of query items from the provided set of URL parameters. Each key-value
+    /// pair of URL parameters corresponds to a query item, where the key is the query item's name
+    /// and the value is the query item's value.
```
Comment (NadzeyaShpakouskaya) [inline Sources/NetworkService/ParametersEncoder/URLParametersEncoder.swift]: I think we can remove this sentence because it describes the implementation. 
`Each key-value pair of URL parameters corresponds to a query item, where the key is the query item's name and the value is the query item's value.`
WDYT?

File: Tests/NetworkServiceTests/URLParametersEncoderTests/QueryItemsTests.swift [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        let expectedQueryItems = [URLQueryItem(name: " ", value: " ")]
+
+        // When
+        let actualQueryItems = URLParametersEncoder.queryItems(from: urlParameters)
+
+        // Then
+        XCTAssertEqual(actualQueryItems, expectedQueryItems)
+    }
+
+    func testReturnsAmountOfQueryItemsThatEqualsToAmountOfURLParameters() throws {
```
Comment (NadzeyaShpakouskaya) [inline Tests/NetworkServiceTests/URLParametersEncoderTests/QueryItemsTests.swift:48]: Why do we need tests for 2 and 3 parameters, it doesn't look like an edge case.
I can understand if you would add cases with 5 or 10 parameters, but 2 and 3 look excessive to me.

I prefer the separate test cases for 2 and 3, 5 or 10 parameters.

**Review by NadzeyaShpakouskaya — CHANGES_REQUESTED** [review]
Please review previous discussions and consider minor improvements:

**Review by iBYarchitect — APPROVED** [review]
LGTM

File: Tests/NetworkServiceTests/URLParametersEncoderTests/QueryItemsTests.swift (hunk 11 above)
Comment (kirpichonok) [reply Tests/NetworkServiceTests/URLParametersEncoderTests/QueryItemsTests.swift:48]: What do 5 and 10 cases give? I wrote cases with 2 and 3 elements to show 2 different values. The case with one element seemed to me not representative enough.
What edge cases do you suggest? 🙂
For 5 and 10 cases, I'll need to create 2 dictionaries with such number of elements. And what will we get for it?

File: Sources/NetworkService/ParametersEncoder/URLParametersEncoder.swift (hunk 10 above)
Comment (kirpichonok) [reply Sources/NetworkService/ParametersEncoder/URLParametersEncoder.swift]: Fixed.

File: Tests/NetworkServiceTests/URLParametersEncoderTests.swift (hunk 3 above)
Comment (kirpichonok) [reply Tests/NetworkServiceTests/URLParametersEncoderTests.swift]: Fixed.

File: Tests/NetworkServiceTests/URLParametersEncoderTests.swift (hunk 4 above)
Comment (kirpichonok) [reply Tests/NetworkServiceTests/URLParametersEncoderTests.swift]: Fixed.

**Review by NadzeyaShpakouskaya — APPROVED** [review]
LGTM
